src/mame/sinclair/scorpion.cpp

```diff
@@ -126,67 +186,106 @@ void scorpion_state::port_1ffd_w(offs_t offset, u8 data)
+			scorpion_update_memory();
 		}
 	}
-	return m_program->read_byte(offset + 0x4000);
+	return m_program.read_byte(offset + 0x4000);
 }
 
 void scorpion_state::scorpion_mem(address_map &map)
 {
-	map(0x0000, 0x3fff).bankr(m_bank_ram[0]).w(FUNC(scorpion_state::spectrum_128_ram_w<0>));
+	map(0x0000, 0x3fff).bankr(m_bank_rom[0]);
 	map(0x0000, 0x3fff).view(m_bank0_rom);
-	m_bank0_rom[0](0x0000, 0x3fff).bankr(m_bank_rom[0]);
+	m_bank0_rom[0](0x0000, 0x3fff).bankr(m_bank_ram[0]).w(FUNC(scorpion_state::spectrum_128_ram_w<0>));
 
 	map(0x4000, 0x7fff).bankr(m_bank_ram[1]).w(FUNC(scorpion_state::spectrum_128_ram_w<1>));
 	map(0x8000, 0xbfff).bankrw(m_bank_ram[2]);
 	map(0xc000, 0xffff).bankr(m_bank_ram[3]).w(FUNC(scorpion_state::spectrum_128_ram_w<3>));
 }
 
-void scorpion_state::scorpion_io(address_map &map)
+void scorpion_state::scorpion_ioext(address_map &map)
 {
 	map.unmap_value_high();
-	map(0x001f, 0x001f).mirror(0xff00).rw(m_beta, FUNC(beta_disk_device::status_r), FUNC(beta_disk_device::command_w));
-	map(0x003f, 0x003f).mirror(0xff00).rw(m_beta, FUNC(beta_disk_device::track_r), FUNC(beta_disk_device::track_w));
-	map(0x005f, 0x005f).mirror(0xff00).rw(m_beta, FUNC(beta_disk_device::sector_r), FUNC(beta_disk_device::sector_w));
-	map(0x007f, 0x007f).mirror(0xff00).rw(m_beta, FUNC(beta_disk_device::data_r), FUNC(beta_disk_device::data_w));
-	map(0x00fe, 0x00fe).select(0xff00).rw(FUNC(scorpion_state::spectrum_ula_r), FUNC(scorpion_state::spectrum_ula_w));
-	map(0x00ff, 0x00ff).mirror(0xff00).rw(m_beta, FUNC(beta_disk_device::state_r), FUNC(beta_disk_device::param_w));
-	map(0x4021, 0x4021).mirror(0x3fdc).w(FUNC(scorpion_state::port_7ffd_w));
-	map(0x8021, 0x8021).mirror(0x3fdc).w("ay8912", FUNC(ay8910_device::data_w));
-	map(0xc021, 0xc021).mirror(0x3fdc).rw("ay8912", FUNC(ay8910_device::data_r), FUNC(ay8910_device::address_w));
-	map(0x0021, 0x0021).mirror(0x3fdc).w(FUNC(scorpion_state::port_1ffd_w));
+	map(0x0022, 0x0022).select(0xffdc) // FE | xxxxxxxxxx1xxx10
+		.rw(FUNC(scorpion_state::spectrum_ula_r), FUNC(scorpion_state::spectrum_ula_w));
+	map(0x0023, 0x0023).mirror(0xffdc) // FF | xxxxxxxxxx1xxx11
+		.r(FUNC(scorpion_state::floating_bus_r));
+	map(0x0021, 0x0021).mirror(0x13fdc) // 1FFD | 00xxxxxxxx1xxx01
+		.w(FUNC(scorpion_state::port_1ffd_w));
+	map(0x4021, 0x4021).mirror(0x13fdc) // 7FFD | 01xxxxxxxx1xxx01
+		.w(FUNC(scorpion_state::port_7ffd_w));
+
+	map(0xa021, 0xa021).mirror(0x1fdc) // BFFD | 101xxxxxxx1xxx01
+		.lw8(NAME([this](u8 data) { m_ay[m_ay_selected]->data_w(data); }));
+	map(0xe021, 0xe021).mirror(0x1fdc) // FFFD | 111xxxxxxx1xxx01
+		.lr8(NAME([this]() { return m_ay[m_ay_selected]->data_r(); })).w(FUNC(scorpion_state::ay_address_w));
+
+	// Mouse
+	map(0xfadf, 0xfadf).lr8(NAME([this]() { return 0x80 | (m_io_mouse[2]->read() & 0x07); }));
+	map(0xfbdf, 0xfbdf).lr8(NAME([this]() { return m_io_mouse[0]->read(); }));
+	map(0xffdf, 0xffdf).lr8(NAME([this]() { return ~m_io_mouse[1]->read(); }));
```

Comment:
Due to C integer promotion rules, the second lambda here won’t return `u8` natively and may end up needing to instantiate an adapter function.  Please force it to return `u8`, either like:
```c++
	map(0xffdf, 0xffdf).lr8(NAME([this] () -> u8 { return ~m_io_mouse[1]->read(); }));
```

or like:
```c++
	map(0xffdf, 0xffdf).lr8(NAME([this] () -> { return u8(~m_io_mouse[1]->read()); }));
```